### PR TITLE
Add support for reserved devices

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area usability
/kind enhancement
/merge squash

**What this PR does / why we need it**:
This PR adds support for reserved EQXM devices. It introduces the new `packet.provider.extensions.gardener.cloud/v1alpha1.WorkerConfig` API which allows end-users to specify the list of reservation IDs as well as a "policy" indicating whether solely those IDs shall be used.

**Special notes for your reviewer**:
✅ This PR depends on a new release of https://github.com/gardener/machine-controller-manager-provider-equinix-metal with support for the `reservationIDs` and `onlyReserved` fields in its provider specification. ~Hence, it's marked as draft for now.~ (see #164).

**Release note**:

```feature user
It is now possible to specify a list of reserved devices per worker pool. For more information please take a look at the [documentation](https://github.com/gardener/gardener-extension-provider-packet/blob/master/docs/usage-as-end-user.md#workerconfig).
```

### CODE REVIEW

File: charts/internal/machineclass/values.yaml [hunk 1]
```diff
@@ -20,3 +20,7 @@ machineClasses:
   credentialsSecretRef:
     name: cloudprovider
     namespace: shoot-namespace
+# reservationIDs:
+# - res1
+# - res2
+# onlyReserved: false
```
Comment (deitch) [inline charts/internal/machineclass/values.yaml:26]: I think `reservedOnly` or maybe even `reservedDevicesOnly` would be better.

Comment (deitch) [reply charts/internal/machineclass/values.yaml:26]: That means changing all of the areas it is produced or consumed, sorry.

**Review by deitch — CHANGES_REQUESTED** [review]
I had a minor syntactical change request (to align with what I think the EQXM folks would want and make it a little easier to read).

Also, can you include an example of it being in a shoot? Even just in docs?

File: charts/internal/machineclass/values.yaml (hunk 1 above)
Comment (rfranzke) [reply charts/internal/machineclass/values.yaml]: No issues, I like it, definitely better :)
Done in 58d2278007312002fb12242173bdcfee68305460.

File: docs/usage-as-end-user.md [hunk 2]
```diff
@@ -61,6 +61,27 @@ apiVersion: packet.provider.extensions.gardener.cloud/v1alpha1
...
+reservationIDs:
+- my-reserved-device-1
+- my-reserved-device-2
+reservedDevicesOnly: false
+```
+
+The `.reservationIDs[]` list contains the list of IDs of the reserved devices.
+The `.reservedDevicesOnly` field indicates whether only reserved devices should be used (based on the list of reservation IDs)
+when new machines are created.
+If it's set to `false` and the list of reservation IDs is exhausted then the next available device (unreserved) will be
```
Comment (deitch) [inline docs/usage-as-end-user.md:82]: I would change the language here to:

```
The `.reservedDevicesOnly` field indicates whether only reserved devices from the provided list of reservation IDs should be used when new machines are created. It always will attempt to create a device from one of the reservation IDs. If none is available, the behaviour depends on the setting:

* `true`: return an error
* `false`: request a regular on-demand device
```

Also, what is the default?

File: docs/usage-as-end-user.md [hunk 3]
```diff
@@ -61,6 +61,27 @@ apiVersion: packet.provider.extensions.gardener.cloud/v1alpha1
...
+used.
+
 ## Example `Shoot` manifest
 
 Please find below an example `Shoot` manifest:
```
Comment (deitch) [inline docs/usage-as-end-user.md:87]: Can we add an example with reservations?

**Review by deitch — CHANGES_REQUESTED** [review]
Documentation things.

File: docs/usage-as-end-user.md (hunk 2 above)
Comment (rfranzke) [reply docs/usage-as-end-user.md]: done in d514ede7cb983e27237ce6301f51ec1004785630

File: docs/usage-as-end-user.md (hunk 3 above)
Comment (rfranzke) [reply docs/usage-as-end-user.md:91]: done in d514ede7cb983e27237ce6301f51ec1004785630

**Review by deitch — APPROVED** [review]
LGTM. Do we need to wait for the MCM update before merging this in?

**Comment by rfranzke** [comment]
Technically, we could also merge it right now, however, it would not work, of course. Though, if you want then we can get it in.

**Comment by deitch** [comment]
Let's wait, then.

**Pull request merged**